sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala

```diff
@@ -359,6 +359,21 @@ trait StateStoreProvider {
   /** Return an instance of [[StateStore]] representing state data of the given version */
   def getStore(version: Long): StateStore
 
+  /**
+   * This is an optional method, used by snapshotStartBatchId option when reading state generated
+   * by join operation as data source.
+   * Return an instance of [[StateStore]] representing state data of the given version.
+   * The State Store will be constructed from the batch at startVersion, and applying delta files
+   * up to the endVersion. If there is no snapshot file of batch startVersion, an exception will
+   * be thrown.
+   *
+   * @param startVersion checkpoint version of the snapshot to start with
+   * @param endVersion checkpoint version to end with
+   */
+  def getStore(startVersion: Long, endVersion: Long): StateStore =
```

Comment:
I wonder whether there is a way (or we should) to expose these APIs only to the state data source reader ? i.e. not make them public apis on the state store provider

cc - @HeartSaVioR - for best practice pattern here

Reply:
One way may be encapsulating these methods in a new trait called `FineGrainedStateDataSource`?